components/butterfly/src/server/expire.rs

```diff
@@ -1,46 +1,53 @@
-//! `Departed`.
+//! `Departed`. Also purge any rumors that have expired.
 
-use crate::{rumor::{RumorKey,
-                    RumorType},
-            server::{timing::Timing,
-                     Server}};
+use crate::server::{timing::Timing,
+                    Server};
+use chrono::offset::Utc;
 use habitat_common::liveliness_checker;
 use std::{thread,
           time::Duration};
 
-const LOOP_DELAY_MS: u64 = 500;
-
 pub fn spawn_thread(name: String, server: Server, timing: Timing) -> std::io::Result<()> {
     thread::Builder::new().name(name)
                           .spawn(move || -> ! { run_loop(&server, &timing) })
                           .map(|_| ())
 }
 
 fn run_loop(server: &Server, timing: &Timing) -> ! {
-    loop {
-        liveliness_checker::mark_thread_alive().and_divergent();
+    habitat_core::env_config_duration!(ExpireThreadSleepMillis, HAB_EXPIRE_THREAD_SLEEP_MS => from_millis, Duration::from_millis(500));
+    let sleep_ms: Duration = ExpireThreadSleepMillis::configured_value().into();
 
-        let newly_confirmed_members =
-            server.member_list
-                  .members_expired_to_confirmed_mlw(timing.suspicion_timeout_duration());
+    habitat_core::env_config_duration!(ExpireThreadPurgeSecs, HAB_EXPIRE_THREAD_PURGE_SECS => from_secs, Duration::from_secs(60));
+    let purge_secs: Duration = ExpireThreadPurgeSecs::configured_value().into();
 
-        for id in newly_confirmed_members {
-            server.rumor_heat
-                  .start_hot_rumor(RumorKey::new(RumorType::Member, &id, ""));
-        }
+    let mut purge_counter = Duration::from_secs(0);
 
-        let newly_departed_members =
-            server.member_list
-                  .members_expired_to_departed_mlw(timing.departure_timeout_duration());
+    loop {
+        liveliness_checker::mark_thread_alive().and_divergent();
 
-        for id in newly_departed_members {
-            server.rumor_heat.purge(&id);
-            server.rumor_heat
-                  .start_hot_rumor(RumorKey::new(RumorType::Member, &id, ""));
+        server.member_list
+              .members_expired_to_confirmed_mlw(timing.suspicion_timeout_duration());
+
+        server.member_list
+              .members_expired_to_departed_mlw(timing.departure_timeout_duration());
+
+        purge_counter += sleep_ms;
+
+        // Rather than trying to do this potentially expensive operation every loop iteration,
+        // let's only do it every once in awhile.
+        if purge_counter >= purge_secs {
+            let now = Utc::now();
+            server.election_store.purge_expired(now);
+            server.update_store.purge_expired(now);
+            server.service_store.purge_expired(now);
+            server.service_config_store.purge_expired(now);
+            server.service_file_store.purge_expired(now);
+            server.member_list.purge_expired_mlw(now);
+            purge_counter = Duration::from_secs(0);
```

Comment:
Sure?  I don't have strong opinions on this.

Reply:
The point of this was the recognition that this expire thread loops every 500ms and we don't need to be expiring rumors with that kind of frequency.  Rather than create a separate thread for expiring rumors, I just added another delay to this loop that's longer than the member expiration delay.  It felt less complicated and resource intensive to do it this way, but I'm happy to split it out into another thread if that's preferable.